IEx Integration

c-brenn opened the pull request:
Added support for IEx shell within vim - requires either Neovim or
ConqueShell. For ConqueShell - the setting `g:ConqueTerm_CloseOnEnd`
should be enabled to properly exit IEx.

There are two new commands:
- `:IEx {command}`
- `:IExHide`

`:IEx` opens a new IEx session if none exists (iex -S mix). If one
already exists, it switches to that window (or reopens the window if it
was closed).
`:IEx` can take a command, which it will run. For example:
`:IEx h Enum.reverse`

`:IEx` also has <tab> completion - using the same completion function as
`:ExDoc` uses.

`:IEx` opens a new horizontal split for the IEx session. The size of the
split can be configured by changing `g:alchemist_iex_term_size`.
Similarly, the horizontal split can be changed to vertical by changing
`g:alchemist_iex_term_split` - either `split` or `vsplit` are valid
values.

`:IExHide` hides the window that the IEx session is in.

c-brenn committed:
IEx Integration

Added support for IEx shell within vim - requires either Neovim or
ConqueShell. For ConqueShell - the setting `g:ConqueTerm_CloseOnEnd`
should be enabled to properly exit IEx.

There are two new commands:
- `:IEx {command}`
- `:IExHide`

`:IEx` opens a new IEx session if none exists (iex -S mix). If one
already exists, it switches to that window (or reopens the window if it
was closed).
`:IEx` can take a command, which it will run. For example:
`:IEx h Enum.reverse`

`:IEx` also has <tab> completion - using the same completion function as
`:ExDoc` uses.

`:IEx` opens a new horizontal split for the IEx session. The size of the
split can be configured by changing `g:alchemist_iex_term_size`.
Similarly, the horizontal split can be changed to vertical by changing
`g:alchemist_iex_term_split` - either `split` or `vsplit` are valid
values.

`:IExHide` hides the window that the IEx session is in.

c-brenn commented:
[Here's a video of the proposed feature in action - in both vim and neovim](https://asciinema.org/a/az7mfhhivm6v67awe0oq59xz8)

At the end you can see the IEx session not closing properly in Vim. This is fixed by setting `g:ConqueTerm_CloseOnEnd` to 1.

Closes #27

slashmili commented:
It's much better than what I was imagining! 

The IEx hide feature is also good. What do you if IEx command is able to toggle the window? First time opens, second time hides. I'm suggesting this because in my config I would like to map it to '<leader>i'. It opens the shell and I'd use the same thing to close it

c-brenn commented:
Nice idea. I think it should hide the buffer if the IEx buffer is the current buffer, otherwise it opens it. Sound good?

c-brenn committed:
Toggle IEx

`:IEx` now hides the IEx session under some circumstances.
If the current buffer is the IEx session, and no command was passed,
`:IEx` will hide the session.

Refactored some of the iex code into helper functions and changed
`-nargs` from `?` to `*`

c-brenn committed:
IEx Buffer check bug

Fixed a bug where `:IEx` was incorrectly checking if the current buffer
was the IEx buffer.

slashmili commented:
If it's current buffer or it's opened in a window.

BTW as you mentioned if IEx supports range it would be really good.

c-brenn commented:
At the moment, it hides IEx if it's the current buffer. Otherwise it opens IEx. From playing around with it, I think this makes the most sense.

I'll take a look at giving it a range, not sure how to best handle it - from what I understand, optional ranges are a bit of a hack (`:h :command-range`).

slashmili commented:
I just tried it and if I am in different buffer/window it jumps to IEx buffer which sounds good.

I'd suggest to make these changes as well:

``` diff
diff --git a/after/plugin/alchemist.vim b/after/plugin/alchemist.vim
index c98e00f..145388f 100644
--- a/after/plugin/alchemist.vim
+++ b/after/plugin/alchemist.vim
@@ -325,10 +325,10 @@ endif
 if !exists('g:alchemist_iex_term_split')
   let g:alchemist_iex_term_split = "split"
 endif
-if exists('g:ConqueTerm_Loaded')
-  let s:alchemist_iex_runner = "ConqueTerm"
-elseif has('nvim')
+if has('nvim')
   let s:alchemist_iex_runner = "terminal"
+elseif exists('g:ConqueTerm_Loaded')
+  let s:alchemist_iex_runner = "ConqueTerm"
 endif

 function! s:iex_open_cmd()
@@ -372,7 +372,11 @@ function! alchemist#open_iex(command)
   else
     " no IEx buffer exists, open a new one
     exec s:iex_open_cmd()
-    exec s:alchemist_iex_runner . " iex -S mix"
+    if !filereadable('mix.exs')
+        exec s:alchemist_iex_runner . " iex"
+    else
+        exec s:alchemist_iex_runner . " iex -S mix"
+    end
     let s:alchemist_iex_buffer = bufnr("%")
     call s:iex_enter_user_command(a:command, '')
   endif
```

c-brenn commented:
Excellent suggestions. I hadn't thought about mot being in a mix project. I'll make those changes in the next few hours

c-brenn committed:
Check for mix.exs & Nvim priority over Conque

Nvim has priority over ConqueShell now.
If no `mix.exs` is available, `iex` is loaded normally.

slashmili commented:
Merged to master.

Thanks again for the good work.

I need to spend some time on documenting the configs before announcing it
